components/eam/src/chemistry/modal_aero/modal_aero_coag.F90

```diff
@@ -1,2337 +1,1363 @@
+!
+! Numerical treatment:
+!  - Note that the updates for different modes are calculated in a sequential manner using a specific 
+!    ordering because
+!    - accum   number loss depends on accum number
+!    - pcarbon number loss depends on pcarbon and accum number
+!
+!  - The average number mixing ratio over current time step
+!    of other modes are used to calculate the number loss of a mode.
+!----------------------------------------------------------------------------------------------------
 
+      real(wp), intent(in) :: ybetaij0(max_coagpair) ! coag rate coefficient
+      real(wp), intent(in) :: ybetaii0(max_coagpair) ! coag rate coefficient
+      real(wp), intent(in) :: ybetajj0(max_coagpair) ! coag rate coefficient
+      real(wp), intent(in) :: deltat
 
-! *** free-molecular form:  equation h.7a of whitby et al. (1991)
+      real(wp), intent(in)    :: qnum_bgn(1:max_mode)  ! beginning values of number mixing ratios
+      real(wp), intent(inout) :: qnum_end(1:max_mode)  ! end values of number mixing ratios
+      real(wp), intent(out)   :: qnum_tavg(1:max_mode) ! time average defined as 0.5*(bgn+end)
 
-         coagfm0 = kfmatac * sqdgat * bm0ij(n1,n2n,n2a) * (   &
-                   esat01 + r * esac01 + two * r2 * esat01 * esac04   &
-                 + rx4 * esat09 * esac16 + ri3 * esat16 * esac09   &
-                 + two * ri1 * esat04 + esac01  )
+      ! local variables
+      real(wp) :: tmpa, rateij
+      real(wp) ::       rateii
 
+      !--------------------------------------------
+      ! accum mode number loss due to self-coag
+      !--------------------------------------------
+      call qnum_update_selfcoag( ybetajj0(1), deltat, qnum_bgn(nacc), qnum_end(nacc) )
 
-! *** loss to accumulation mode
+      qnum_tavg(nacc) = (qnum_bgn(nacc) + qnum_end(nacc))*0.5_wp
 
-! *** harmonic mean
+      !----------------------------------------------------------------------------
+      ! pcarbon mode number loss - approximate analytical solution 
+      ! using average number conc. for accum mode
+      !----------------------------------------------------------------------------
+      rateij = max( 0.0_wp, deltat*ybetaij0(2)*qnum_tavg(nacc) )
+      rateii = max( 0.0_wp, deltat*ybetaii0(2) )
 
-      coagatac0 = coagnc0 * coagfm0 / ( coagnc0 + coagfm0 )
+      call qnum_update_self_and_intermodal_coag( rateij, rateii, qnum_bgn(npca), qnum_end(npca) )
 
-      qn12 = coagatac0
+      qnum_tavg(npca) = (qnum_bgn(npca) + qnum_end(npca))*0.5_wp
 
+      !-----------------------------------------------------------------------------------------
+      ! aitken mode number loss - approximate analytical solution
+      ! using average number conc. for accum, pcarbon, and marine-org accum modes
+      !-----------------------------------------------------------------------------------------
+      tmpa = ybetaij0(1)*qnum_tavg(nacc)
+      tmpa = tmpa + ybetaij0(3)*qnum_tavg(npca)
```

Comment:
Can we rename `tmpa` to something more meaningful? It is okay to use `tmp` but can add a string like `tmp_avg` or something similar?

Reply:
I changed `tmpa` to `bijqnumj` in [this commit](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/0aacb82f99405c7f4207680de523988fb31ce395) and explained the meaning in the comments: it's the intermodal coagulation coefficient bij multiplied by the number mixing ratio of mode j. 

Some of the variables in these low-level subroutines are hard to give good names to, because they correspond to mathematical expressions that are specific to the numerical algorithms in use and they don't a clear physical meaning. By using the variable name `bijqnumj` for what used to be `tmpa`, I'm trying to spell out what the quantity being calculated is, which can be useful for readers who want to check or understand the numerical algorithm. Hope this is acceptable.